subprojects/execution/src/main/java/org/gradle/internal/execution/steps/CacheStep.java

```diff
@@ -128,18 +133,20 @@ public UnitOfWork getWork() {
             });
     }
 
-    private static void cleanupTreesAfterLoadFailure(UnitOfWork work) {
-        work.visitOutputTrees((name, type, root) -> {
-            try {
-                if (root.exists()) {
-                    if (root.isDirectory()) {
-                        FileUtils.cleanDirectory(root);
-                    } else {
-                        FileUtils.forceDelete(root);
+    private static void cleanOutputsAfterLoadFailure(UnitOfWork work) {
+        work.visitOutputProperties((name, type, roots) -> {
```

Comment:
Why don't you visit the output trees? That seems to be better for cacheable tasks.

Reply:
I'd like to use the `UnitOfWork` here, and basically divorce it from `CacheableEntity` eventually. They should map to the same properties in the end anyway.